Panic: runtime error: invalid memory address or nil pointer dereference

Hi there,

please note that this is an issue tracker reserved for bug reports and feature requests.

For general questions please use the gitter channel or the Ethereum stack exchange at https://ethereum.stackexchange.com.

#### System information

Geth version: 1.8.12

OS & Version: OSX Mojave

Branch, Commit Hash or Release: 2.2.1

#### Actual behaviour
Panic observed in quorum while adding a new peer using raft.addPeer

#### Steps to reproduce the behaviour
1. Start a quorum container with its own enode url in static-nodes and permissioned-nodes
2. Update its static-nodes and permissioned-nodes with the enode url of the other peer which needs to be added to the network
3. Restart the first node and start the other node
4. From the first node, perform raft.addPeer("enode url of the other node")
5. Panic can be observed on the first node

#### Backtrace

````
2019-02-15 12:03:54,471 DEBG 'geth' stdout output:
2019-02-15 12:03:54.471571 E | rafthttp: failed to find member 2 in cluster 1000

2019-02-15 12:03:54,532 DEBG 'geth' stdout output:
2019-02-15 12:03:54.531005 E | rafthttp: failed to find member 2 in cluster 1000

2019-02-15 12:03:54,575 DEBG 'geth' stdout output:
2019-02-15 12:03:54.575198 E | rafthttp: failed to find member 2 in cluster 1000

2019-02-15 12:03:54,634 DEBG 'geth' stdout output:
2019-02-15 12:03:54.634313 E | rafthttp: failed to find member 2 in cluster 1000

2019-02-15 12:03:54,666 DEBG 'geth' stdout output:
INFO [02-15|12:03:54] ignoring expected ConfChangeAddNode for initial peer raft id=2
INFO [02-15|12:03:54] start snapshot                           applied index=26 last snapshot index=1

2019-02-15 12:03:54,673 DEBG 'geth' stdout output:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd59554]

goroutine 73 [running]:
github.com/ethereum/go-ethereum/raft.(*ProtocolManager).buildSnapshot(0xc4200cb1e0, 0x0)
	/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/raft/snapshot.go:56 +0x314
github.com/ethereum/go-ethereum/raft.(*ProtocolManager).triggerSnapshot(0xc4200cb1e0, 0x1b)
	/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/raft/snapshot.go:83 +0x18f
github.com/ethereum/go-ethereum/raft.(*ProtocolManager).eventLoop(0xc4200cb1e0)
	/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/raft/handler.go:802 +0x743
created by github.com/ethereum/go-ethereum/raft.(*ProtocolManager).startRaft
	/quorum/build/_workspace/src/github.com/ethereum/go-ethereum/raft/handler.go:496 +0x789

2019-02-15 12:03:54,692 DEBG fd 10 closed, stopped monitoring <POutputDispatcher at 139734393637056 for <Subprocess at 139734393883032 with name geth in state RUNNING> (stdout)>
2019-02-15 12:03:54,693 INFO exited: geth (exit status 2; expected)
2019-02-15 12:03:54,693 DEBG received SIGCLD indicating a child quit
2019-02-15 12:03:54,894 DEBG 'privacy_manager' stdout output:
12:03:53 [INFO] Finished synchronization with https://172.13.2.2:9001/
12:03:53 [INFO] Starting synchronization with http://172.13.2.2:9001/
12:03:53 [INFO] Finished synchronization with http://172.13.2.2:9001/
12:03:53 [INFO] Starting synchronization with https://172.13.1.2:9001/
12:03:53 [DEBUG] Request from 172.13.1.2:57448: ApiPartyInfo (PartyInfo {piUrl = "http://172.13.1.2:9001/", piRcpts = fromList [("p6m2ZLsWrwUZ40LfkwCI38abGd5SVFIwiKQnrbcOCD4=","http://172.13.1.2:9001/"),("vXyr0gidIQbE3ZpBThlqEKhIc2C48a5n62ftfD4ypUY=","http://172.13.2.2:9001/")], piParties = fromList ["https://172.13.2.2:9001/","http://172.13.2.2:9001/","http://172.13.1.2:9001/","https://172.13.1.2:9001/"]}); Response: ApiPartyInfoR (PartyInfo {piUrl = "http://172.13.1.2:9001/", piRcpts = fromList [("p6m2ZLsWrwUZ40LfkwCI38abGd5SVFIwiKQnrbcOCD4=","http://172.13.1.2:9001/"),("vXyr0gidIQbE3ZpBThlqEKhIc2C48a5n62ftfD4ypUY=","http://172.13.2.2:9001/")], piParties = fromList ["https://172.13.2.2:9001/","http://172.13.2.2:9001/","http://172.13.1.2:9001/","https://172.13.1.2:9001/"]})
12:03:53 [INFO] Finished synchronization with https://172.13.1.2:9001/

2019-02-15 12:04:03,902 DEBG 'privacy_manager' stdout output:
12:04:03 [INFO] Starting synchronization with https://172.13.2.2:9001/

2019-02-15 12:04:04,900 DEBG 'privacy_manager' stdout output:
12:04:03 [INFO] Finished synchronization with https://172.13.2.2:9001/
12:04:03 [INFO] Starting synchronization with http://172.13.2.2:9001/
12:04:03 [INFO] Finished synchronization with http://172.13.2.2:9001/
12:04:03 [INFO] Starting synchronization with https://172.13.1.2:9001/
12:04:03 [DEBUG] Request from 172.13.1.2:57448: ApiPartyInfo (PartyInfo {piUrl = "http://172.13.1.2:9001/", piRcpts = fromList [("p6m2ZLsWrwUZ40LfkwCI38abGd5SVFIwiKQnrbcOCD4=","http://172.13.1.2:9001/"),("vXyr0gidIQbE3ZpBThlqEKhIc2C48a5n62ftfD4ypUY=","http://172.13.2.2:9001/")], piParties = fromList ["https://172.13.2.2:9001/","http://172.13.2.2:9001/","http://172.13.1.2:9001/","https://172.13.1.2:9001/"]}); Response: ApiPartyInfoR (PartyInfo {piUrl = "http://172.13.1.2:9001/", piRcpts = fromList [("p6m2ZLsWrwUZ40LfkwCI38abGd5SVFIwiKQnrbcOCD4=","http://172.13.1.2:9001/"),("vXyr0gidIQbE3ZpBThlqEKhIc2C48a5n62ftfD4ypUY=","http://172.13.2.2:9001/")], piParties = fromList ["https://172.13.2.2:9001/","http://172.13.2.2:9001/","http://172.13.1.2:9001/","https://172.13.1.2:9001/"]})
12:04:03 [INFO] Finished synchronization with https://172.13.1.2:9001/

````

--- a/raft/handler.go
+++ b/raft/handler.go
@@ -754,7 +754,7 @@ func (pm *ProtocolManager) eventLoop() {
 					case raftpb.ConfChangeAddNode:
 						if pm.isRaftIdRemoved(raftId) {
 							log.Info("ignoring ConfChangeAddNode for permanently-removed peer", "raft id", raftId)
-						} else if raftId <= uint16(len(pm.bootstrapNodes)) {
+						} else if peer := pm.peers[raftId]; peer != nil && raftId <= uint16(len(pm.bootstrapNodes))  {
 							// See initial cluster logic in startRaft() for more information.
 							log.Info("ignoring expected ConfChangeAddNode for initial peer", "raft id", raftId)
 

